qualtran/bloqs/controlled_state_preparation.ipynb

```diff
@@ -0,0 +1,240 @@
+    "circuit = bb.finalize(state=state)\n",
+    "\n",
+    "show_bloq(circuit)\n",
+    "coefficients = circuit.tensor_contract()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "And finally a comparison of the results obtained with the original state used."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "accuracy = np.dot(coefficients, np.array(state_coefs).conj())\n",
+    "\n",
+    "print(f\"original state used: {tuple(state_coefs)}\")\n",
+    "print(f\"circuit result:      {tuple(coefficients)}\\n\")\n",
+    "print(f\"accuracy: {abs(accuracy)}\\n\")\n",
+    "\n",
+    "print(\"Comparison (coefficients in polar form):\")\n",
+    "for i, (c, s) in enumerate(zip(coefficients, state_coefs)):\n",
+    "  print(f\"  |{i:0{n_qubits}b}> result: {round(abs(c),4)} ∠{round(np.angle(c, deg=True),2)}º  \"+\\\n",
+    "        f\"exact: {round(abs(s),4)} ∠{round(np.angle(s, deg=True),2)}º\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Using the adjoint\n",
+    "\n",
+    "This block also implement the adjoint, that is, preparing $|0\\rangle$ from a given $|\\psi\\rangle$. Following an equivalent scheme to the previous example, we prepare and un-prepare a state using the adjoint $U^\\dagger$ of the preparation gate $U$:\n",
+    "$$\n",
+    "  |0\\rangle = U^\\dagger |\\psi\\rangle = U^\\dagger U |0\\rangle\n",
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "qsp_adj = ControlledStatePreparationUsingRotations(\n",
+    "    n_qubits=n_qubits, rot_reg_size=rot_reg_size, state=tuple(state_coefs), adjoint=True\n",
+    ")\n",
+    "\n",
+    "bb = BloqBuilder()\n",
+    "control = bb.add(OneState())\n",
+    "state = bb.join(np.array([bb.add(ZeroState()) for _ in range(n_qubits)]))\n",
+    "phase_gradient = bb.add(PhaseGradientState(rot_reg_size))\n",
+    "control, state, phase_gradient = bb.add(\n",
+    "    qsp, control=control, target_state=state, phase_gradient=phase_gradient\n",
+    ")\n",
+    "control, state, phase_gradient = bb.add(\n",
+    "    qsp_adj, control=control, target_state=state, phase_gradient=phase_gradient\n",
+    ")\n",
+    "bb.add(OneEffect(), q=control)\n",
+    "bb.add(PhaseGradientState(bitsize=rot_reg_size, adjoint=True), phase_grad=phase_gradient)\n",
+    "circuit = bb.finalize(state=state)\n",
+    "\n",
+    "show_bloq(circuit)\n",
+    "coefficients = circuit.tensor_contract()\n",
+    "\n",
+    "accuracy = coefficients[0] # <coefficients|0> = coefficients[0]\n",
+    "print(f\"accuracy: {abs(accuracy)}\\n\")\n",
+    "\n",
+    "print(\"Coefficients in polar form:\")\n",
+    "for i, c in enumerate(coefficients):\n",
+    "  print(f\"  |{i:0{n_qubits}b}> result: {round(abs(c),4)} ∠{round(np.angle(c, deg=True),2)}º\")"
```

Comment:
Why are the "angle" values here so odd? I get

Coefficients in polar form:
  |000> result: 1.0 ∠0.0º
  |001> result: 0.0 ∠-115.57º
  |010> result: 0.0 ∠-64.32º
  |011> result: 0.0 ∠57.47º
  |100> result: 0.0 ∠-53.75º
  |101> result: 0.0 ∠17.61º
  |110> result: 0.0 ∠37.75º
  |111> result: 0.0 ∠-150.2º

Reply:
It is normal, because the amplitude of the complex value is zero (or a value of the order of 1e-16) I think `np.angle` just outputs some random value.

Reply:
right, you have a bunch of zeros with different relative signs. I think adding

``` python
# zero out small coefficients
coefficients[np.where(abs(coefficients) < 1e-12)] = 0.0
```
right before your print is probably a good idea. Also it looks like you're using 2 spaces instead of 4 in the notebook, I guess black doesn't check the notebooks.